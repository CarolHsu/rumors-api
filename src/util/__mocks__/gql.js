/**
 * Usage:
 *
 * jest.mock('path-to-gql/gql');
 *
 * it('some test', async () => {
 *   gql.__addMockResponse({data: {queryName: result})
 *
 *   const result = await functionUnderTestThatInvokesGql();
 *
 *   expect(gql.__getRequests).toMatchSnapshot(); // assert request sent
 *   expect(result).toMatchSnapshot();
 *
 *   gql.__reset() // Must reset to reset mock responses & call sequence number!
 * })
 */

let seq = 0;
let mockResponses = [];
let requests = [];

function gql(query, ...substitutions) {
  if (!mockResponses[seq]) {
    throw Error(
      `gql Mock error: No response found for request #${seq}. Please add mock response first.`
    );
  }

  return variables => {
    const queryAndVariable = {
      query: String.raw(query, ...substitutions),
    };

    if (variables) queryAndVariable.variables = variables;

    requests.push(queryAndVariable);

    return Promise.resolve(mockResponses[seq++]);
  };
}

gql.__addMockResponse = resp => mockResponses.push(resp);
gql.__getRequests = () => requests;
gql.__reset = () => {
  seq = 0;
  mockResponses = [];
  requests = [];
};

export default gql;
