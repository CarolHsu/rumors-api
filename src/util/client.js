import config from 'config';
import elasticsearch from 'elasticsearch';

export default new elasticsearch.Client({
  host: config.get('ELASTICSEARCH_URL'),
  log: 'trace',
});
