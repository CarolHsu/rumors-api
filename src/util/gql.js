import fetch from 'node-fetch';
import rollbar from '../rollbarInstance';
import url from 'url';
import config from 'config';

const API_URL = config.get('URL_RESOLVER_URL') || 'http://localhost:4000/';

// Usage:
//
// import gql from './util/GraphQL';
// gql`query($var: Type) { foo }`({var: 123}).then(...)
//
// gql`...`() returns a promise that resolves to immutable Map({data, errors}).
//
// We use template string here so that Atom's language-babel does syntax highlight
// for us.
//
// GraphQL Protocol: http://dev.apollodata.com/tools/graphql-server/requests.html
//
export default (query, ...substitutions) => (variables, search) => {
  const queryAndVariable = {
    query: String.raw(query, ...substitutions),
  };

  if (variables) queryAndVariable.variables = variables;

  let status;
  const URL = `${API_URL}${url.format({ query: search })}`;

  return fetch(URL, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(queryAndVariable),
  })
    .then(r => {
      status = r.status;
      return r.json();
    })
    .then(resp => {
      if (status === 400) {
        throw new Error(
          `GraphQL Error: ${resp.errors
            .map(({ message }) => message)
            .join('\n')}`
        );
      }
      if (resp.errors) {
        // When status is 200 but have error, just print them out.
        // eslint-disable-next-line no-console
        console.error('GraphQL operation contains error:', resp.errors);
        rollbar.error(
          'GraphQL error',
          {
            body: JSON.stringify(queryAndVariable),
            url: URL,
          },
          { resp }
        );
      }
      return resp;
    });
};
