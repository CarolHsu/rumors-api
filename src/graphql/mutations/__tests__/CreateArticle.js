import gql from 'util/GraphQL';
import client from 'util/client';

import MockDate from 'mockdate';

describe('CreateArticle', () => {
  it('creates articles', async () => {
    MockDate.set(1485593157011);
    const { data, errors } = await gql`
      mutation(
        $text: String!
        $reference: ArticleReferenceInput!
      ) {
        CreateArticle(
          text: $text
          reference: $reference
        ) {
          id
        }
      }
    `({
      text: 'FOO FOO', reference: { type: 'LINE' },
    }, { userId: 'test', from: 'foo' });
    MockDate.reset();

    expect(errors).toBeUndefined();

    const doc = await client.get({ index: 'articles', type: 'basic', id: data.CreateArticle.id });
    delete doc._id; // delete auto-generated id from being snapshot

    expect(doc).toMatchSnapshot();

    // Cleanup
    await client.delete({ index: 'articles', type: 'basic', id: data.CreateArticle.id });
  });
});
