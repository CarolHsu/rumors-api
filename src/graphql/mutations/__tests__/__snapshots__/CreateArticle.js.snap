exports[`CreateArticle creates articles 1`] = `
Object {
  "_index": "articles",
  "_source": Object {
    "createdAt": "2017-01-28T08:45:57.011Z",
    "from": "foo",
    "references": Array [
      Object {
        "type": "LINE",
      },
    ],
    "replyConnectionIds": Array [],
    "text": "FOO FOO",
    "updatedAt": "2017-01-28T08:45:57.011Z",
    "userId": "test",
  },
  "_type": "basic",
  "_version": 1,
  "found": true,
}
`;
